tensorflow/compiler/mlir/xla/transforms/hlo_outline_fusion.cc

```diff
@@ -0,0 +1,557 @@
+  SmallVector<Value, 4> inputs;
+  DenseSet<Value> input_set;
+  std::unordered_set<Operation*> op_set(pattern.begin(), pattern.end());
+
+  for (Operation* op : pattern) {
+    for (Value operand : op->getOperands()) {
+      Operation* operand_op = operand.getDefiningOp();
+      if (op_set.find(operand_op) != op_set.end()) {
+        // skip if defining op is in the pattern
+        continue;
+      }
+      if (input_set.insert(operand).second) {
+        inputs.push_back(operand);
+      }
+    }
+  }
+  return std::move(inputs);
+}
+
+SmallVector<Value, 4> GetOutputsOfFusionPattern(const FusionPattern& pattern) {
+  SmallVector<Value, 4> outputs;
+  DenseSet<Value> output_set;
+  std::unordered_set<Operation*> op_set(pattern.begin(), pattern.end());
+  for (Operation* op : pattern) {
+    for (Value result : op->getResults()) {
+      bool has_external_user = false;
+      for (auto& use : result.getUses()) {
+        if (op_set.find(use.getOwner()) == op_set.end()) {
+          has_external_user = true;
+          break;
+        }
+      }
+      if (has_external_user) {
+        if (output_set.insert(result).second) {
+          outputs.push_back(result);
+        }
+      }
+    }
+  }
+  return outputs;
+}
+
+SmallVector<Value, 4> GetInputusAndOutputsOfOp(Operation* op) {
+  SmallVector<Value, 4> ios;
+  ios.reserve(op->getNumOperands() + op->getNumResults());
+  for (Value operand : op->getOperands()) {
+    ios.push_back(operand);
+  }
+  for (Value result : op->getResults()) {
+    ios.push_back(result);
+  }
+  return ios;
+}
+
+FusionPattern MergeFusionPattern(
+    const FusionPattern& lhs, const FusionPattern& rhs) {
+  FusionPattern pattern(lhs);
+  pattern.insert(pattern.end(), rhs.begin(), rhs.end());
+  return pattern;
+}
+
+class ShapeConstraintAnalysis {
+ public:
+   ShapeConstraintAnalysis(const SmallVector<Operation*, 4>& op_list)
+       : op_list_(op_list) {
+     InitState();
+     PropagateEquality();
+   }
+
+   bool HasSameShape(Value lhs, Value rhs) {
+     return (shape_checker_.count(lhs) &&
+         shape_checker_.count(rhs) &&
+         shape_checker_[lhs].Get() == shape_checker_[rhs].Get());
+   }
+
+ private:
+  void InitState() {
+    for (auto op : op_list_) {
+      for (auto v : GetInputusAndOutputsOfOp(op)) {
+        shape_checker_[v].Get() = v;
+      }
+    }
+  }
+
+  void PropagateEquality() {
+    bool converged = true;
+    do {
+      converged = true;
+      for (auto op : op_list_) {
+        if (IsElementWise(op)) {
+          auto ios = GetInputusAndOutputsOfOp(op);
+          for (auto v : ios) {
+            Value ref = ios[0];
+            if (shape_checker_[ref].Get() != shape_checker_[v].Get()) {
+              converged = false;
+              shape_checker_[ref].Merge(&shape_checker_[v]);
+            }
+          }
+        }
+      }
+    } while (!converged);
+  }
+
+  SmallVector<Operation*, 4> op_list_;
```

Comment:
Seems like this is only needed to initialize the state and you don't need this copy at all